### PR TITLE
AP_Mount:Add GIMBAL_MANAGER_SET_ATTITUDE support

### PR DESCRIPTION
**Pull request by khanasif786**
Task from Gimbal enhancement list #20985  
consumes [GIMBAL_MANAGER_SET_ATTITUDE](https://mavlink.io/en/messages/common.html#GIMBAL_MANAGER_SET_ATTITUDE) messages

This needs to be merged before it
[#308](https://github.com/ArduPilot/mavlink/pull/308)

### CODE REVIEW

File: libraries/AP_Mount/AP_Mount.cpp [hunk 1]
```diff
@@ -337,6 +337,68 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
...
+        roll_deg = RadiansToCentiDegrees(roll_rad)*100.0f;
+        pitch_deg = RadiansToCentiDegrees(pitch_rad)*100.0f;
+        yaw_deg = RadiansToCentiDegrees(yaw_rad)*100.0f;
+        backend->set_angle_target(roll_deg, pitch_deg, yaw_deg, flags & GIMBAL_MANAGER_FLAGS_YAW_LOCK);
+        return;
+    }
+
+
+    if (!isnan(packet.angular_velocity_x) && !isnan(packet.angular_velocity_y) && !isnan(packet.angular_velocity_y)) {
+        const float roll_rate_degs = RadiansToCentiDegrees(packet.angular_velocity_x)/100.0f;
```
Comment (amilcarlucas) [inline libraries/AP_Mount/AP_Mount.cpp:395]: can you multiply by 0.01f instead?

**Review by amilcarlucas — CHANGES_REQUESTED** [review]

File: libraries/GCS_MAVLink/GCS_Common.cpp [hunk 2]
```diff
@@ -3870,6 +3870,9 @@ void GCS_MAVLINK::handle_common_message(const mavlink_message_t &msg)
     case MAVLINK_MSG_ID_GIMBAL_DEVICE_ATTITUDE_STATUS:
         handle_mount_message(msg);
         break;
```
Comment (peterbarker) [inline libraries/GCS_MAVLink/GCS_Common.cpp:3872]:
```suggestion
```

File: libraries/AP_Mount/AP_Mount.cpp [hunk 3]
```diff
@@ -337,6 +337,68 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
...
+    } else {
+        backend = get_instance(instance - 1);
+    }
+
+    if (backend == nullptr) {
+        return;
+    }
+
+    // check flags for change to RETRACT
+    uint32_t flags = (uint32_t)packet.flags;
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount.cpp:360]:
```suggestion
    const uint32_t flags = (uint32_t)packet.flags;
```

Why the cast?

Is it worth pulling this out of the packet or just using the field twice where needed?

File: libraries/AP_Mount/AP_Mount.cpp [hunk 4]
```diff
@@ -337,6 +337,68 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
...
+        return;
+    }
+
+    if (!isnan(packet.q[0]) && !isnan(packet.q[1]) && !isnan(packet.q[2]) && !isnan(packet.q[3])) {
+        // convert quaternion to euler angles
+        Quaternion att_quat;
+        att_quat.q1 = packet.q[0];
+        att_quat.q2 = packet.q[1];
+        att_quat.q3 = packet.q[2];
+        att_quat.q4 = packet.q[3];
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount.cpp:378]:
```suggestion
   const Quaternion att_quat{packet.q};
    if (!q.isnan()) {
        // convert quaternion to euler angles
```

File: libraries/AP_Mount/AP_Mount.cpp [hunk 5]
```diff
@@ -337,6 +337,68 @@ MAV_RESULT AP_Mount::handle_command_do_gimbal_manager_pitchyaw(const mavlink_com
...
+
+        // quaternion to eular conversion
+        float roll_rad, pitch_rad, yaw_rad;
+        att_quat.to_euler(roll_rad, pitch_rad, yaw_rad);
+
+        // radian to deg conversion
+        float roll_deg, pitch_deg, yaw_deg;
+        roll_deg = RadiansToCentiDegrees(roll_rad)*100.0f;
+        pitch_deg = RadiansToCentiDegrees(pitch_rad)*100.0f;
+        yaw_deg = RadiansToCentiDegrees(yaw_rad)*100.0f;
```
Comment (peterbarker) [inline libraries/AP_Mount/AP_Mount.cpp:388]:
```suggestion
        const float roll_deg = degrees(roll_rad);
        const float pitch_deg = degrees(pitch_rad);
        const float yaw_deg = degrees(yaw_rad);
```

similarly on the angular rates

**Review by peterbarker — CHANGES_REQUESTED** [review]
Seem about right.

File: libraries/AP_Mount/AP_Mount.cpp (hunk 1 above)
Comment (khanasif786) [reply libraries/AP_Mount/AP_Mount.cpp:395]: @amilcarlucas Thanks a lot. Will remember this in future.

**Comment by khanasif786** [comment]
@peterbarker @amilcarlucas Thanks for the suggestions.

**Comment by khanasif786** [comment]
Sorry. I was thinking we can request review from two people at the same time. Please ignore this.

**Review by rmackay9 — APPROVED** [review]
Looks good to me.  After this is merged we should add a new item to the developer gimbal control page too.  https://ardupilot.org/dev/docs/mavlink-gimbal-mount.html.  Not critical of course though.

**Comment by khanasif786** [comment]
@rmackay9 okay will make a pull on wiki also afterward merging.

**Comment by peterbarker** [comment]
> Looks good to me. After this is merged we should add a new item to the developer gimbal control page too. https://ardupilot.org/dev/docs/mavlink-gimbal-mount.html. Not critical of course though.

I have a concern here that we are not treating this mavlink message in the same way that we treat the vehicle `SET_ATTITUDE_TARGET` message.  @lthall went to a lot of trouble to unscramble that egg for us to we ended up with correct behaviour there.  It would be a shame to put another scrambled egg in.

The basic difference between this implementation and the vehicle handling of `SET_ATTITUDE_TARGET`  is that with the vehicle message you are not specifying the rate at which to come to the new attitude, you are specifying the rate at which the vehicle should be moving when it gets to that attitude.

**Comment by peterbarker** [comment]
I've merged the mavlink part of this.

**Comment by peterbarker** [comment]
I've tested this on a T3v3 and it does control the gimbal.

The attitude always takes precedent over the rate, so if your quaternion is valid you will come to that attitude as fast as the gimbal allows (pretty damned fast in the T3 case!)

If you have an invalid quaternion then it honours the rates (I have not checked the rate directions).

These were the MAVProxy commands I was using to test:
```
message GIMBAL_MANAGER_SET_ATTITUDE 0 0 0 0 [float("NAN"),float("NAN"),float("NAN"),float("NAN")] 0.01 0.01 0.01

message GIMBAL_MANAGER_SET_ATTITUDE 0 0 0 0 [1,0,0,0] 0.01 0.01 0.01

message GIMBAL_MANAGER_SET_ATTITUDE 0 0 0 0 [1,0.1,0.1,0.1] 0.001 0.001 0.001
```

**Comment by peterbarker** [comment]
I suggest we ignore (and probably throw a warning periodically) if we ever receive a SET_GIMBAL_ATTITUDE message that has both the attitude and the rate set.  We don't sensibly handle that case right now, so it's a reasonable precaution and will allow us to implement the same semantics as the vehicle equivalent at some stage.

**Comment by rmackay9** [comment]
@peterbarker,

Re callers who provide both rate and angle controls, I think that for now at least the correct behaviour is what we have in this PR which is to consume the angle targets and ignore the rate targets.  There are no gimbals that support accepting both at the same time so it's a bit different than what Leonard did with our attitude controllers.

**Comment by rmackay9** [comment]
@peterbarker thanks very much for the review by the way!

**Comment by rmackay9** [comment]
@peterbarker,

OK, I don't think it's necessary but it's a team effort so maybe you can provide a suggestion on those few extra lines.

**Comment by rmackay9** [comment]
I've rebased this on master after merging the mavlink submodule update.

**Comment by peterbarker** [comment]
> @peterbarker,
> 
> OK, I don't think it's necessary but it's a team effort so maybe you can provide a suggestion on those few extra lines.

https://github.com/peterbarker/ardupilot/pull/new/pr/no-att-and-rate-together

**Comment by lthall** [comment]
You could reject messages that provide both angle and rate. That way we wouldn't find ourselves in a situation where we start supporting both and break someone's system.

**Comment by peterbarker** [comment]
@khanasif786 did you want to cherry-pick my suggested change?

**Comment by rmackay9** [comment]
Looks like PeterB's suggestion had a little typo in it.  So might need to fix that somehow.

Maybe it is "is_nan()" instead of "isn_nan()"?

**Comment by peterbarker** [comment]
> Looks like PeterB's suggestion had a little typo in it. So might need to fix that somehow.
> 
> Maybe it is "is_nan()" instead of "isn_nan()"?

Fixed.

**Comment by peterbarker** [comment]
@rmackay9 if you're happy I'm happy now :-)

**Comment by khanasif786** [comment]
@peterbarker @rmackay9 can it be merged?

**Comment by peterbarker** [comment]
> @peterbarker @rmackay9 can it be merged?

I've marked it for DevCall to make sure there's a deadline it gets discussed for merging.  Randy may choose to merge earlier.

**Comment by peterbarker** [comment]
Merged, thanks!

**Pull request merged**